### ISSUE TITLE
Types of property '_ils' are incompatible.

### ISSUE THREAD

**Issue by download13** [staltz/xstream#316]
When `merge`ing streams of component types of a [discriminating union](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions) the internal listener seems to be trying to assign a variable of the compound type to a variable of a component type.

You can see in the [example repository](https://github.com/download13/xstream_type_bug) I made that typechecking produces the following error:

```
index.ts:17:7 - error TS2322: Type 'Stream<FirstComponentType>' is not assignable to type 'Stream<CompoundType>'.
  Types of property '_ils' are incompatible.
    Type 'InternalListener<FirstComponentType>[]' is not assignable to type 'InternalListener<CompoundType>[]'.
      Type 'InternalListener<FirstComponentType>' is not assignable to type 'InternalListener<CompoundType>'.
        Type 'CompoundType' is not assignable to type 'FirstComponentType'.
          Type 'SecondComponentType' is not assignable to type 'FirstComponentType'.
```

Basically, it seems like `merge` won't output types compatible with the `CompountType` even when all of it's input stream are of component types. It works if merge contains all of the component types however. You can test it by adding `second$` to the `xs.merge` call.

Is this intentional for a reason I don't understand, or is it a bug? I ran into it while converting DOM events into `Action`s to feed into my reducer function in a cyclejs app.

**Comment by whitecolor** [comment]
```ts
type Type1 = number
type Type2 = string

type CompoundType = Type1 | Type2;

const first: Type1 = 1;
const second: Type2 = '1';

const first$ = xs.of(first);

const x$: Stream<CompoundType> = first$; // this gives an error

```

This is due `strictFunctionTypes` enabled.

But probably it still should be considered as buggy behavior and as can be avoided. I see no reason why `Stream<number>` should not be actually not assignable to `Stream<number | string>`.

With `xstream` it is caused by having visible internal props on the class with `InternalListener<T>`
```ts
protected _ils: Array<InternalListener<T>>;
protected _dl: InternalListener<T>;
```

@staltz ?

**Comment by whitecolor** [comment]
I made a PR for this https://github.com/staltz/xstream/pull/317
